evi-react-native-example/modules/audio/ios/SoundPlayer.swift

```diff
@@ -0,0 +1,102 @@
+import ExpoModulesCore
+import AVFoundation
+import Foundation
+
+// This is a simple sound player that uses the AVAudioPlayer API to supports playing a single audio
+// clip at a time. Maintaining a queue of pending audio clips is left to the user (to be done
+// in Javascript).
+public class SoundPlayer {
+    private var lastAudioPlayer: AVAudioPlayer?
+    public func stopPlayback() {
+        self.lastAudioPlayer?.stop()
+        self.lastAudioPlayer = nil
+    }
+
+    // This method takes a chunk of base64-encoded audio and plays it via the AVAudioPlayer API.
+    // It uses `withUnsafeThrowingContinuation` so that on the Javascript side it will return a promise
+    // that resolves when the audio is finished playing.
+    public func playBase64Audio(_ base64String: String) async throws {
+        logAudioSessionInfo()
+        // It's important to define `delegate` in this scope, as `audioPlayer.delegate` is a weak reference and this will be deallocated if it is only
+        // defined within the inner scope of `withUnsafeThrowingContinuation`.
+        var delegate: ContinuationDelegate!
+        var audioPlayer: AVAudioPlayer!
+        return try await withUnsafeThrowingContinuation { continuation in
+            do {
+                guard let data = Data(base64Encoded: base64String) else {
+                    continuation.resume(throwing: NSError(domain: "AudioModule", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 string"]))
```

Comment:
if you wanna make this more swifty, you can define a custom error for this domain. and similarly for the Microphone. then you can just throw `SoundPlayerError.invalidData`. `NSError` is a legacy api from obj-c days

```
enum SoundPlayerError: Error {
    case invalidData
    case audioError
    case someOtherErrorWithDetails(extraDeets: String)
}
```

Reply:
Done in [Avoid NSError, create bespoke error types](https://github.com/HumeAI/hume-api-examples/pull/100/commits/d306e959f36cf8707e4c6387ad29df5a95a474c7)!